### ISSUE TITLE
coinbasepro orderBook depth issue in ccxt.pro

### ISSUE THREAD

**Issue by dmnode** [ccxt/ccxt#8164]
Hi,
I have a problem with coinbasepro watchOrderBook.
When I run simple code below (developed earlier for different testing purpose, but can be used now), node thread is using 100% of CPU, which for my complex code leads to failure as my software runs out of resources. When I replace coinbasepro.js with an older version, where

```
this.orderbooks[symbol] = this.orderBook ({});
```
is replaced by
```
const depth = 50; // default depth is 50
this.orderbooks[symbol] = this.orderBook ({}, depth);
```
that code is consuming some 37% of CPU and all is fine.
Thanks for checking.

- OS: Ubuntu
- Programming Language version: node v12.20.0
- CCXT version: ccxt.pro@0.5.3

```

'use strict';

const ccxtpro = require ('ccxt.pro')
    , log = require ('ololog').handleNodeErrors ().unlimited
    , ansi = require ('ansicolor').nice
    , fs = require ('fs')
    , exchange = new ccxtpro.coinbasepro ({ enableRateLimit: true, rateLimit: 1000 })
    , symbol = 'BTC/USD'
    , limit = 10
    , threshold = 100

;(async () => {

    let lastFetchPrice
    let lastFetched

    async function watchOrderBook (exchange, symbol, limit) {
        let counter = 0
        let lastWatchPrice
        while (true) {

            try {

                const orderbook = await exchange.watchOrderBook (symbol, limit)
                const price = orderbook.bids[0][0]
                counter = (price === lastWatchPrice) ? ++counter : 0
                lastWatchPrice = price;

                const now = exchange.milliseconds ()
                const watchTimestamp = new Date (now)
                const fetchTimestamp = new Date (lastFetched)

                fs.appendFileSync ('coinbaseproTips.dat',`${watchTimestamp} ${lastWatchPrice} ${lastFetchPrice} ${fetchTimestamp}\n`);

                if ((counter > threshold) && (lastWatchPrice !== lastFetchPrice)) {
                    log.red (exchange.iso8601 (now).dim, 'frozen tip detected:', lastWatchPrice, lastFetchPrice, exchange.iso8601 (lastFetched).dim)
                    fs.appendFileSync ('frozenTips.dat', `${watchTimestamp} ${lastWatchPrice} ${lastFetchPrice} ${fetchTimestamp}\n`)

                } else {
                    log.green (exchange.iso8601 (now).dim, price, lastFetchPrice, exchange.iso8601 (lastFetched).dim)
                }
            } catch (e) {
                log.red (e)
                throw e
            }
        }
    }

    async function fetchOrderBook (exchange, symbol, limit) {
        while (true) {
            try {
                const orderbook = await exchange.fetchOrderBook (symbol, limit)
                lastFetchPrice = orderbook.bids[0][0]
                lastFetched = new Date (exchange.last_response_headers['Date']).getTime ()
            } catch (e) {
                log.red (e)
            }
        }
    }

    await Promise.all ([
        watchOrderBook (exchange, symbol, limit),
        fetchOrderBook (exchange, symbol, limit),
    ])

}) ()

```

**Comment by kroitor** [comment]
Hi!

Let's start with removing all irrelevant parts from your code to make it a short snippet to work with. In short, those synchronous writes to disk kinda kill all the benefits of WebSockets, this is not the way you want to design a truly reactive and fast application – the writes to disk or to network have to be asynchronous as well.

Let's start with this short snippet, how much % CPU does it consume?

```JavaScript
const ccxtpro = require ('./ccxt.pro')

console.log ('CCXT Pro Version:', ccxtpro.version)

const exchange = new ccxtpro.coinbasepro ({
    'enableRateLimit': true,
})

const symbol = 'BTC/USD'
const limit = 10

;(async () => {

    while (true) {
        try {
            const orderbook = await exchange.watchOrderBook (symbol, limit)
            console.log (orderbook.asks[0], orderbook.bids[0])
        } catch (e) {
            console.log (e.constructor.name, e.message)
            process.exit ()
        }
    }

}) ()
```

**Comment by dmnode** [comment]
Hi Igor,
Thanks, but even with this simple code the same problem. Old coinbasepro.js in ccxt.pro = 25% CPU, new coinbasepro.js in ccxt.pro = 109.6% CPU, as measured by top. In my real-word case I run more asynchronous tasks, the code gets stuck in coinbasepro and fails.

NEW CCXT.PRO
VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
892648 152724  31408 R 109.6   1.9   0:34.82 node 

OLD CCXT.PRO
VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
895368  89156  31372 S  25.0   1.1   0:02.80 node

**Comment by kroitor** [comment]
@dmnode ok, we will add a workaround in coinbasepro.js and will get back to you shortly.

**Comment by kroitor** [comment]
@dmnode alright, i've pushed a fix, can you please upgrade to 0.5.4+ and report back if you still have issues with it?
